vortex-ipc/src/messages/decoder.rs

```diff
@@ -0,0 +1,220 @@
+
+use bytes::{Buf, BytesMut};
+use flatbuffers::{root, root_unchecked, Follow};
+use itertools::Itertools;
+use vortex_array::{flatbuffers as fba, ArrayData, Context};
+use vortex_buffer::Buffer;
+use vortex_dtype::DType;
+use vortex_error::{vortex_bail, vortex_err, VortexExpect, VortexResult};
+use vortex_flatbuffers::message as fb;
+use vortex_flatbuffers::message::{MessageHeader, MessageVersion};
+
+/// A message decoded from an IPC stream.
+///
+/// Note that the `Array` variant cannot fully decode into an [`ArrayData`] without a [`Context`]
+/// and a [`DType`]. As such, we partially decode into an [`ArrayParts`] and allow the caller to
+/// finish the decoding.
+#[derive(Debug)]
+pub enum DecoderMessage {
+    Array(ArrayParts),
+    Buffer(Buffer),
+    DType(DType),
+}
+
+/// ArrayParts represents a partially decoded Vortex array.
+/// It can be completely decoded calling `into_array_data` with a context and dtype.
+pub struct ArrayParts {
+    row_count: usize,
+    // Typed as fb::Array
+    array_flatbuffer: Buffer,
+    array_flatbuffer_loc: usize,
+    buffers: Vec<Buffer>,
+}
+
+impl Debug for ArrayParts {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("ArrayComponents")
+            .field("row_count", &self.row_count)
+            .field("array_flatbuffer", &self.array_flatbuffer.len())
+            .field("buffers", &self.buffers.len())
+            .finish()
+    }
+}
+
+impl ArrayParts {
+    pub fn into_array_data(self, ctx: Arc<Context>, dtype: DType) -> VortexResult<ArrayData> {
+        ArrayData::try_new_viewed(
+            ctx,
+            dtype,
+            self.row_count,
+            self.array_flatbuffer,
+            // SAFETY: ArrayComponents guarantees the buffers are valid.
+            |buf| unsafe { Ok(fba::Array::follow(buf, self.array_flatbuffer_loc)) },
+            self.buffers,
+        )
+    }
+}
+
+#[derive(Default)]
+enum State {
+    #[default]
+    Length,
+    Header(usize),
+    Array(ReadingArray),
+    Buffer(ReadingBuffer),
+}
+
+struct ReadingArray {
+    header: Buffer,
+    buffers_length: usize,
+}
+
+struct ReadingBuffer {
+    length: usize,
+    length_with_padding: usize,
+}
+
+pub enum NextMessage {
+    Some(DecoderMessage),
+    NeedMore(usize),
+}
+
+/// A stateful reader for decoding IPC messages from an arbitrary stream of bytes.
+#[derive(Default)]
+pub struct MessageDecoder {
+    state: State,
+}
+
+impl MessageDecoder {
+    /// Attempt to read the next message from the bytes object.
+    /// If the message is incomplete, the function will return `NeedMore` with the _total_ number
+    /// of bytes needed. The next call to read_next _should_ provide at least this number of bytes.
+    pub fn read_next(&mut self, bytes: &mut BytesMut) -> VortexResult<NextMessage> {
+        match &self.state {
+            State::Length => {
+                if bytes.len() < 4 {
+                    return Ok(NextMessage::NeedMore(4));
```

Comment:
This should be `4 - bytes.len()`, if the goal is that `NeedMore(x)` means "I need x more bytes of data to be read from the stream"

Reply:
If we want the semantic to be something else we should document it. In general pub tuple structs should be documented b/c otherwise you need to read a lot of code to figure out what they mean

Reply:
I can move the doc-string to the enum. But the semantic is the number of total bytes for the next successful read.